{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pickle\n",
    "from sklearn.metrics.pairwise import cosine_similarity\n",
    "from tensorflow.keras.models import load_model\n",
    "import joblib\n",
    "import tensorflow as tf\n",
    "from flask import Flask, request, jsonify"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def search_courses(query, top_n=5):\n",
    "    # Load pre-trained model\n",
    "    model = load_model(\"model_recommender.keras\")\n",
    "\n",
    "    # Load preprocessed assets\n",
    "    tfidf = pickle.load(open(\"tfidf_vectorizer.pkl\", \"rb\"))  # Load the fitted vectorizer\n",
    "    course_embeddings = np.load(\"course_embeddings.npy\")\n",
    "    courses = pd.read_csv(\"udemy_courses (1).csv\")\n",
    "\n",
    "    # Transform the query using the fitted vectorizer\n",
    "    query_vec = tfidf.transform([query]).toarray()\n",
    "    query_emb = model.predict(query_vec)\n",
    "\n",
    "    # Compute similarities\n",
    "    similarities = cosine_similarity(query_emb, course_embeddings).flatten()\n",
    "    top_indices = similarities.argsort()[::-1][:top_n]\n",
    "\n",
    "    return courses.iloc[top_indices].to_dict(orient=\"records\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * Serving Flask app '__main__'\n",
      " * Debug mode: off\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.\n",
      " * Running on http://127.0.0.1:5000\n",
      "Press CTRL+C to quit\n",
      "127.0.0.1 - - [08/May/2025 01:00:52] \"POST /recommend-by-id HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'int'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[2025-05-08 01:05:20,549] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:20] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:21,089] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:21] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:21,668] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:21] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:22,312] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:22] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:22,839] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:22] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:23,382] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:23] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:24,011] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:24] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:24,675] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:24] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:25,416] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:25] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:26,173] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:26] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:26,951] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:26] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:27,615] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:27] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:28,292] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:28] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:28,973] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:28] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:29,713] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:29] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:30,441] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:30] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:31,174] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:31] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:31,838] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:31] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:32,520] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:32] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:33,233] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:33] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:34,971] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:34] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:35,605] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:35] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:36,308] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:36] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:37,131] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:37] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:37,780] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:37] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:38,429] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:38] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:39,135] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:39] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:39,911] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:39] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:40,644] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:40] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:41,383] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:41] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:42,010] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:42] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:42,685] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:42] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:43,349] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:43] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:44,086] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:44] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:44,950] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:44] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:45,720] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:45] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:46,406] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:46] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:47,177] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:47] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:47,951] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:47] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:48,705] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:48] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:54,307] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:54] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:54,956] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:54] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:55,572] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:55] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:56,246] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:56] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:56,806] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:56] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:57,506] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:57] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:58,223] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:58] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:58,907] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:58] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:05:59,430] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:05:59] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:06:00,168] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:06:00] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:06:00,792] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:06:00] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:06:01,432] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:06:01] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:06:01,971] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:06:01] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:06:02,628] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:06:02] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:06:03,281] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:06:03] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:06:03,840] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:06:03] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:06:04,432] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:06:04] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:06:05,035] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:06:05] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:06:05,619] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:06:05] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:06:06,383] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:06:06] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:06:08,095] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:06:08] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:06:08,625] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:06:08] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:06:09,185] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:06:09] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:06:09,723] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:06:09] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:06:10,359] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:06:10] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:06:10,977] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:06:10] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:06:11,691] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:06:11] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:06:12,470] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:06:12] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:06:13,196] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:06:13] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:06:13,958] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:06:13] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:06:14,662] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:06:14] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:06:15,395] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:06:15] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:06:16,156] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:06:16] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:06:16,747] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:06:16] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:06:17,324] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:06:17] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:06:17,906] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:06:17] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:06:18,436] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:06:18] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:06:19,146] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:06:19] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:06:19,788] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:06:19] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:06:20,687] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:06:20] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "[2025-05-08 01:06:22,456] ERROR in app: Exception on /recommend-by-id [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 2190, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1486, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1484, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\moham\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\flask\\app.py\", line 1469, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\moham\\AppData\\Local\\Temp\\ipykernel_2852\\3494463846.py\", line 34, in recommend_by_id\n",
      "    course_id = int(course_id)\n",
      "                ^^^^^^^^^^^^^^\n",
      "ValueError: invalid literal for int() with base 10: 'aEpmZUh0ZEJJU1BDWkp3eUNmVFBHdz09'\n",
      "127.0.0.1 - - [08/May/2025 01:06:22] \"POST /recommend-by-id HTTP/1.1\" 500 -\n",
      "127.0.0.1 - - [08/May/2025 01:07:16] \"POST /recommend-by-id HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'int'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [08/May/2025 01:07:34] \"POST /recommend-by-id HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'int'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [08/May/2025 01:08:10] \"POST /recommend-by-id HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'int'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [08/May/2025 01:08:16] \"POST /recommend-by-id HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'int'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [08/May/2025 01:10:31] \"POST /recommend-by-id HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'int'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [08/May/2025 01:11:55] \"POST /recommend-by-id HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'int'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [08/May/2025 01:12:10] \"POST /recommend-by-id HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'int'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [08/May/2025 01:12:33] \"POST /recommend-by-id HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'int'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [08/May/2025 01:14:04] \"POST /recommend-by-id HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'int'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [08/May/2025 01:14:23] \"POST /recommend-by-id HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'int'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [08/May/2025 01:15:02] \"POST /recommend-by-id HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'int'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [08/May/2025 01:15:11] \"POST /recommend-by-id HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'int'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [08/May/2025 01:15:40] \"POST /recommend-by-id HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'int'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [08/May/2025 01:16:46] \"POST /recommend-by-id HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'int'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [08/May/2025 01:16:55] \"POST /recommend-by-id HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'int'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [08/May/2025 01:17:00] \"POST /recommend-by-id HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'int'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [08/May/2025 01:18:15] \"POST /recommend-by-id HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'int'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [08/May/2025 01:18:33] \"POST /recommend-by-id HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'int'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [08/May/2025 01:19:35] \"POST /recommend-by-id HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'int'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [08/May/2025 01:20:21] \"POST /recommend-by-id HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'int'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [08/May/2025 01:20:32] \"POST /recommend-by-id HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'int'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [08/May/2025 01:20:38] \"POST /recommend-by-id HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'int'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [08/May/2025 01:21:36] \"POST /recommend-by-id HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'int'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [08/May/2025 01:22:47] \"POST /recommend-by-id HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'int'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [08/May/2025 01:24:41] \"POST /recommend-by-id HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'int'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [08/May/2025 01:24:56] \"POST /recommend-by-id HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'int'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [08/May/2025 01:25:12] \"POST /recommend-by-id HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'int'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [08/May/2025 01:25:13] \"POST /recommend-by-id HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'int'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [08/May/2025 01:25:14] \"POST /recommend-by-id HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'int'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [08/May/2025 01:25:23] \"POST /recommend-by-id HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'int'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [08/May/2025 01:26:00] \"POST /recommend-by-id HTTP/1.1\" 200 -\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'int'>\n"
     ]
    }
   ],
   "source": [
    "\n",
    "app = Flask(__name__)\n",
    "\n",
    "# Load assets once at startup\n",
    "model = load_model(\"model_recommender.keras\")\n",
    "tfidf = pickle.load(open(\"tfidf_vectorizer.pkl\", \"rb\"))\n",
    "course_embeddings = np.load(\"course_embeddings.npy\")\n",
    "courses = pd.read_csv(\"udemy_courses (1).csv\")\n",
    "encoder_model = tf.keras.models.load_model('course_encoder_model.keras')\n",
    "\n",
    "# === Define recommendation function ===\n",
    "def recommend_related(course_id, courses_df, embeddings, top_n=10):\n",
    "    print(type(course_id))\n",
    "    if course_id not in courses_df['id'].values:\n",
    "        raise ValueError(f\"Course ID {course_id} not found in dataset.\")\n",
    "\n",
    "    idx = courses_df[courses_df['id'] == course_id].index[0]\n",
    "    course_vec = embeddings[idx].reshape(1, -1)\n",
    "\n",
    "    similarities = cosine_similarity(course_vec, embeddings).flatten()\n",
    "    courses_df['similarity'] = similarities\n",
    "\n",
    "    recommendations = courses_df[courses_df['id'] != course_id] \\\n",
    "        .sort_values(by='similarity', ascending=False) \\\n",
    "        .head(top_n)\n",
    "\n",
    "    return recommendations['id'].tolist() \n",
    "    # return recommendations[['id', 'title', 'similarity']]\n",
    "\n",
    "# === Flask route that returns JSON ===\n",
    "@app.route('/recommend-by-id', methods=['POST'])\n",
    "def recommend_by_id():\n",
    "    data = request.get_json()\n",
    "    course_id = data.get(\"course_id\")\n",
    "    course_id = int(course_id)\n",
    "    if course_id is None:\n",
    "        return jsonify({\"error\": \"Missing course_id\"}), 400\n",
    "\n",
    "    try:\n",
    "        recs = recommend_related(course_id, courses, course_embeddings)\n",
    "        return jsonify(recs)  # Return the list directly as JSON\n",
    "    except ValueError as e:\n",
    "        return jsonify({\"error\": str(e)}), 404\n",
    "\n",
    "@app.route('/recommend', methods=['POST'])\n",
    "def recommend():\n",
    "    data = request.get_json()\n",
    "    query = data.get(\"query\")\n",
    "\n",
    "    if not query:\n",
    "        return jsonify({\"error\": \"Missing query\"}), 400\n",
    "\n",
    "    query_vec = tfidf.transform([query]).toarray()\n",
    "    query_emb = model.predict(query_vec)\n",
    "\n",
    "    similarities = cosine_similarity(query_emb, course_embeddings).flatten()\n",
    "    top_indices = similarities.argsort()[::-1][:5]\n",
    "\n",
    "    recommended_courses = courses.iloc[top_indices].to_dict(orient=\"records\")\n",
    "    return jsonify(recommended_courses)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    app.run(port=5000)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "755       695952\n",
       "71       3792262\n",
       "9486     3466516\n",
       "2077      364384\n",
       "51615    1910784\n",
       "Name: id, dtype: int64"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "recommend_related(221674, courses, course_embeddings, top_n=5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### return ids \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * Serving Flask app '__main__'\n",
      " * Debug mode: off\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.\n",
      " * Running on http://127.0.0.1:5000\n",
      "Press CTRL+C to quit\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[1m1/1\u001b[0m \u001b[32m━━━━━━━━━━━━━━━━━━━━\u001b[0m\u001b[37m\u001b[0m \u001b[1m0s\u001b[0m 97ms/step\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [07/May/2025 20:14:43] \"POST /recommend HTTP/1.1\" 200 -\n"
     ]
    }
   ],
   "source": [
    "from flask import Flask, request, jsonify\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pickle\n",
    "from sklearn.metrics.pairwise import cosine_similarity\n",
    "from tensorflow.keras.models import load_model\n",
    "\n",
    "app = Flask(__name__)\n",
    "\n",
    "# Load assets once at startup\n",
    "model = load_model(\"model_recommender.keras\")\n",
    "tfidf = pickle.load(open(\"tfidf_vectorizer.pkl\", \"rb\"))\n",
    "course_embeddings = np.load(\"course_embeddings.npy\")\n",
    "courses = pd.read_csv(\"udemy_courses (1).csv\")\n",
    "\n",
    "@app.route('/recommend', methods=['POST'])\n",
    "def recommend():\n",
    "    data = request.get_json()\n",
    "    query = data.get(\"query\")\n",
    "\n",
    "    if not query:\n",
    "        return jsonify({\"error\": \"Missing query\"}), 400\n",
    "\n",
    "    query_vec = tfidf.transform([query]).toarray()\n",
    "    query_emb = model.predict(query_vec)\n",
    "\n",
    "    similarities = cosine_similarity(query_emb, course_embeddings).flatten()\n",
    "    top_indices = similarities.argsort()[::-1][:10]\n",
    "\n",
    "    # Get only course IDs of the recommended courses\n",
    "    recommended_course_ids = courses.iloc[top_indices]['id'].tolist()\n",
    "\n",
    "    return jsonify(recommended_course_ids)\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    app.run(port=5000)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
